Grid system component

yuki24 opened the pull request:
### Screenshot:

![screen shot 2017-02-27 at 6 18 41 pm](https://cloud.githubusercontent.com/assets/386234/23384498/3912dcce-fd19-11e6-98d3-db44888d33ae.png)

### Interface

```jsx
const boxStyle = {
  backgroundColor: '#f8f8f8'
  height: '100px',
  lineHeight: '100px',
  textAlign: 'center'
}

<div style={{margin: '20px'}}>
  <Row>
    <Col><div style={boxStyle}>1</div></Col>
    <Col><div style={boxStyle}>2</div></Col>
    <Col><div style={boxStyle}>3</div></Col>
    <Col><div style={boxStyle}>4</div></Col>
    <Col><div style={boxStyle}>5</div></Col>
    <Col><div style={boxStyle}>6</div></Col>
  </Row>

  <Row>
    <Col><div style={boxStyle}>1</div></Col>
    <Col><div style={boxStyle}>2</div></Col>
    <Col><div style={boxStyle}>3</div></Col>
    <Col><div style={boxStyle}>4</div></Col>
  </Row>

  <Row>
    <Col><div style={boxStyle}>1</div></Col>
    <Col><div style={boxStyle}>2</div></Col>
    <Col><div style={boxStyle}>3</div></Col>
  </Row>

  <Row>
    <Col><div style={boxStyle}>1</div></Col>
    <Col><div style={boxStyle}>2</div></Col>
  </Row>

  <Row>
    <Col><div style={boxStyle}>1</div></Col>
  </Row>
</div>
```

alloy commented on src/components/grids.tsx:
```diff
@@ -0,0 +1,25 @@
+  padding-right: ${gapSize / 2}px;
+  padding-left: ${gapSize / 2}px;
+  flex-basis: ${props => (100 / numberOfColumns) * (props.size)}%;
+  max-width: ${props => (100 / numberOfColumns) * (props.size)}%;
```
Rather than doing math here and having the user of the API having to define `props.size`, I wonder if we could just use `flex-grow` here? https://css-tricks.com/snippets/css/a-guide-to-flexbox/#article-header-id-10

The size prop can still exist, it would just default to 1 so the user doesn’t need to specify it if they want equal distribution.

yuki24 replied:
It seems like the value of `flex-grow` is relative to other columns while the current grid system has the fixed scale of 12 in total. We could also just use `flex` as suggested here: https://facebook.github.io/react-native/docs/height-and-width.html, but I'm not sure what it'd look like on modern browsers.

I feel like we should experiment with `flex` or other flex-related properties before we get too far so we can make grid perfectly re-usable across multiple platforms.

What are the browser versions we are targeting? Also, is there any example iOS app that actually uses reaction-force that we can use for acceptance?

alloy replied:
> It seems like the value of `flex-grow` is relative to other columns while the current grid system has the fixed scale of 12 in total.

I’m thinking that maybe less static numbers of columns is good for responsive pages (now that Force and microgravity have merged) as well as indeed share with React Native?

> I feel like we should experiment with `flex` or other flex-related properties before we get too far so we can make grid perfectly re-usable across multiple platforms.

I don’t think we need to engineer it too much right now, I think we should aim for the simplest solution that works well for what we need _right now_, so I’d suggest you take the time to make the best use of flexbox (and e.g. get rid of manual calculations).

> What are the browser versions we are targeting?

Good question. @craigspaeth can you say what historically has been the target?

> Also, is there any example iOS app that actually uses reaction-force that we can use for acceptance?

Nah, not yet. Don’t worry right now about these components working or not for iOS. We’ll cross that bridge when we come to it, just make sure to keep it simple so that it will be flexible to adjust to when we _do_ try to make it cross platform.

yuki24 replied:
It seems like `flex` is just the shorthand for `flex-grow`, `flex-shrink` and `flex-basis`, so I'm actually :+1: with just using `flex-grow` anywhere. I'll also remove the `max-width`. My only concern now is browser compatibility (and what I mean by browser is, obviously, IE11).

alloy commented:
Looking good thus far! Nice and short ✨ 

Let’s rename the module to (singular) `grid`, though, as there is only 1.

yuki24 commented:
updated and rebased. Now `size: 1` is the default so you even don't have to specify anything if all the elements should have the same width.

There is one possible problem though. If you specify `size: 2` for example, the width of the column wouldn't equal to the sum of the two widths (see the screenshot and the code below). This is happening because the width of the longer one doesn't include padding values.

![grid](https://cloud.githubusercontent.com/assets/386234/23435608/e0572d3e-fdd5-11e6-90af-6ef4da1d6305.png)

```jsx
<Row>
  <Col><div style={boxStyle}>1</div></Col>
  <Col><div style={boxStyle}>2</div></Col>
  <Col><div style={boxStyle}>3</div></Col>
  <Col><div style={boxStyle}>4</div></Col>
</Row>

<Row>
  <Col size={2}><div style={boxStyle}>1</div></Col>
  <Col><div style={boxStyle}>2</div></Col>
  <Col><div style={boxStyle}>3</div></Col>
</Row>
```

l2succes commented:
@yuki24 I think It'd be worthwhile to implement a responsive grid using flexbox. For one we can then design components that render well in multiple screen sizes including mobile web and even react-native which also supports flexbox.

We can use this project as inspiration for desiging the grid.
http://roylee0704.github.io/react-flexbox-grid/


![css-flexbox-small](https://cloud.githubusercontent.com/assets/296775/23468705/80a5b12c-fe6f-11e6-9dce-f2fcfb47354c.gif)

yuki24 commented:
@l2succes I actually picked up some of the code from flexbox-grid originally (without responsive part though).

yuki24 commented:
Oooh, I totally misunderstood you @l2succes! Yes, we can totally pull in `react-flexbox-grid`.

What do you think @alloy?

alloy commented:
I generally prefer less dependencies.

We really only need a simple squares grid for now, we don’t even need them to be able to span multiple columns, so let’s keep it as simple as possible and is actually required.

![grid](https://cloud.githubusercontent.com/assets/2320/23522397/1e8149c4-ff83-11e6-8462-7faa135a5b23.jpg)

yuki24 commented:
I see this grid on the top page of force.

![screen shot 2017-03-02 at 2 06 20 pm](https://cloud.githubusercontent.com/assets/386234/23522566/74787334-ff51-11e6-92f7-b4865dcce6ef.png)

If even this was out of scope, why would we even need a style guide?

yuki24 commented:
Also, how do our designers build out a page? Do they use a 12 grid system? If yes, we should align with them and implement the same type of grid system.

mzikherman commented:
Are we getting a little ahead of ourselves here? If the purpose of this work is a 'spike' in experimenting how we would build the loyalty page, which is relatively simple, I would not want to spend too much time on super generic stuff before we really understood _how_ and _if_ we are actually going to use them.

alloy commented:
@yuki24 Yeah the point here is to test out using React by building something as simple as the loyalty program registration, we do not need to build out more than we need for just that.

l2succes commented:
Yeah good point @mzikherman. we don't need a very flexible grid system since the layout for the loyalty page is pretty straightforward. @yuki24 Let's hold off on that and just get this basic grid implementation merged in.

yuki24 commented:
@mzikherman just had a conversation with @l2succes on this - and I misunderstood the entire goal of this approach.

I think we can merge this in since the current implementation only uses `flex-grow`.

xtina-starr commented:
👍

alloy commented:
Sounds good 👍 I resolved the merge conflict for you, but the CI fail is legit.

xtina-starr commented:
What up with CI? It appears as if there is a lint error but it isn't showing what the error is.

yuki24 commented:
rebased and fixed all the lint errors 👍

alloy commented:
Aight 👍
